File: cpp/tests/test_local_sim.cpp
#include <cmath>
#include <sstream>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/core/planner.hpp"
#include "navigation/core/trace.hpp"
#include "navigation/core/types.hpp"
#include "navigation/local_planning/geometry.hpp"
#include "navigation/local_planning/simulation.hpp"
#include "navigation/maps/occupancy_grid.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Capability;
using core::LocalTask;
using core::ObstacleLocalPlanner;
using core::ObstacleQuery;
using core::ParamSet;
using core::Pose;
using core::RobotState;
using core::TraceRecorder;
using core::VelocityCommand;
using local_planning::SimConfig;
using local_planning::SimResult;
using local_planning::SimStatus;
using local_planning::simulate;

namespace {

// --- minimal script planners: fixtures that live inside the test file, not
// production algorithms, so they relax no visibility on the real planners -----

// Constant forward speed, no turning -- exercises integrate_unicycle's straight
// branch (omega == 0) and the REACHED / COLLISION / TIMEOUT termination paths.
class StraightPlanner : public ObstacleLocalPlanner {
 public:
  explicit StraightPlanner(double v) : ObstacleLocalPlanner(ParamSet{}), v_(v) {}
  std::string name() const override { return "straight_script"; }
  std::set<Capability> required_capabilities() const override {
    return {Capability::OBSTACLE_QUERY};
  }
  VelocityCommand compute_command(ObstacleQuery&, const RobotState&, const LocalTask&, double,
                                  TraceRecorder*) override {
    return VelocityCommand{v_, 0.0};
  }

 private:
  double v_;
};

// Constant (v, omega) every tick -- the executed trajectory must lie exactly on
// the closed-form arc's circle (radius v/omega).
class ConstantArcPlanner : public ObstacleLocalPlanner {
 public:
  ConstantArcPlanner(double v, double omega) : ObstacleLocalPlanner(ParamSet{}), v_(v), omega_(omega) {}
  std::string name() const override { return "arc_script"; }
  std::set<Capability> required_capabilities() const override {
    return {Capability::OBSTACLE_QUERY};
  }
  VelocityCommand compute_command(ObstacleQuery&, const RobotState&, const LocalTask&, double,
                                  TraceRecorder*) override {
    return VelocityCommand{v_, omega_};
  }

 private:
  double v_;
  double omega_;
};

// Never moves -- net displacement over any window is exactly 0, so this drives
// the STALLED path deterministically.
class ZeroPlanner : public ObstacleLocalPlanner {
 public:
  ZeroPlanner() : ObstacleLocalPlanner(ParamSet{}) {}
  std::string name() const override { return "zero_script"; }
  std::set<Capability> required_capabilities() const override {
    return {Capability::OBSTACLE_QUERY};
  }
  VelocityCommand compute_command(ObstacleQuery&, const RobotState&, const LocalTask&, double,
                                  TraceRecorder*) override {
    return VelocityCommand{0.0, 0.0};
  }
};

// Turns for the first `turn_ticks_` calls (mutating `ticks_`), then drives
// straight. Used to prove simulate() calls planner.reset() at episode start:
// re-running the SAME instance twice must not let the first episode's counter
// leak into the second.
class CountingPlanner : public ObstacleLocalPlanner {
 public:
  explicit CountingPlanner(int turn_ticks)
      : ObstacleLocalPlanner(ParamSet{}), turn_ticks_(turn_ticks) {}
  std::string name() const override { return "counting_script"; }
  std::set<Capability> required_capabilities() const override {
    return {Capability::OBSTACLE_QUERY};
  }
  void reset() override { ticks_ = 0; }
  VelocityCommand compute_command(ObstacleQuery&, const RobotState&, const LocalTask&, double,
                                  TraceRecorder*) override {
    bool turning = ticks_ < turn_ticks_;
    ++ticks_;
    return turning ? VelocityCommand{0.3, 0.8} : VelocityCommand{0.5, 0.0};
  }

 private:
  int turn_ticks_;
  int ticks_ = 0;
};

// Large open grid (no interior obstacles) so straight/arc runs never collide
// unless the test deliberately places a wall.
maps::OccupancyGrid2D open_grid(int n, double resolution = 0.5) {
  std::vector<bool> free_cells(static_cast<size_t>(n) * n, true);
  return maps::OccupancyGrid2D(n, n, resolution, 0.0, 0.0, free_cells);
}

// A conservative default SimConfig: stall_window larger than any test's
// max_steps so STALLED never fires as a side effect in tests that aren't
// exercising it.
SimConfig base_config(double control_dt, int max_steps, double goal_tolerance,
                      double footprint_radius = 0.1) {
  return SimConfig{control_dt, max_steps, goal_tolerance, footprint_radius, max_steps + 1, 1e-9};
}

}  // namespace

// ① straight command -> REACHED, time_to_goal == steps * dt --------------------

TEST(LocalSim, StraightReachesGoalWithExactTimeToGoal) {
  auto grid = open_grid(40);
  RobotState start{Pose{2.0, 2.0, 0.0}, 0.0, 0.0};  // clear of the map boundary
  LocalTask task{Pose{3.0, 2.0, 0.0}, {}};
  SimConfig config = base_config(/*dt=*/0.1, /*max_steps=*/50, /*goal_tolerance=*/0.05);
  StraightPlanner planner(/*v=*/0.5);

  SimResult r = simulate(planner, grid, start, task, config, nullptr);

  ASSERT_EQ(r.status, SimStatus::REACHED);
  EXPECT_TRUE(r.success);
  EXPECT_EQ(r.steps, 20);  // 0.5 m/s * 0.1 s * 20 ticks = 1.0 m == goal distance
  EXPECT_DOUBLE_EQ(r.time_to_goal, r.steps * config.control_dt);
}

// ② driving straight into a wall -> COLLISION within a couple of ticks ---------

TEST(LocalSim, WallAheadCausesCollision) {
  std::vector<std::string> rows(8, std::string(10, '.'));
  for (auto& row : rows) row[3] = '#';  // vertical wall at col 3 -> x in [1.5, 2.0)
  auto grid = test::make_grid(rows);
  RobotState start{Pose{1.2, 1.0, 0.0}, 0.0, 0.0};
  LocalTask task{Pose{4.0, 1.0, 0.0}, {}};
  SimConfig config = base_config(/*dt=*/0.1, /*max_steps=*/50, /*goal_tolerance=*/0.01,
                                 /*footprint_radius=*/0.15);
  StraightPlanner planner(/*v=*/1.0);

  SimResult r = simulate(planner, grid, start, task, config, nullptr);

  ASSERT_EQ(r.status, SimStatus::COLLISION);
  EXPECT_FALSE(r.success);
  EXPECT_LE(r.steps, 3);
  // The colliding tick's pose is never appended to the executed trajectory.
  EXPECT_EQ(r.trajectory.size(), static_cast<size_t>(r.steps));
}

// ③ zero command sustained -> STALLED -------------------------------------------

TEST(LocalSim, ZeroCommandStalls) {
  auto grid = open_grid(40);
  RobotState start{Pose{2.0, 2.0, 0.0}, 0.0, 0.0};
  LocalTask task{Pose{15.0, 15.0, 0.0}, {}};
  SimConfig config{/*control_dt=*/0.1, /*max_steps=*/50, /*goal_tolerance=*/0.05,
                   /*footprint_radius=*/0.1, /*stall_window=*/3, /*stall_distance=*/0.01};
  ZeroPlanner planner;

  SimResult r = simulate(planner, grid, start, task, config, nullptr);

  ASSERT_EQ(r.status, SimStatus::STALLED);
  EXPECT_FALSE(r.success);
  EXPECT_EQ(r.steps, config.stall_window);  // displacement is 0 from tick 1 on
  EXPECT_DOUBLE_EQ(r.distance_traveled, 0.0);
}

// ④ goal unreachable within the budget -> TIMEOUT --------------------------------

TEST(LocalSim, ExhaustsBudgetWithoutReachingOrStalling) {
  auto grid = open_grid(200, /*resolution=*/1.0);  // 200m x 200m, plenty of room
  RobotState start{Pose{100.0, 100.0, 0.0}, 0.0, 0.0};
  LocalTask task{Pose{-500.0, -500.0, 0.0}, {}};  // far outside the map -> unreachable
  SimConfig config = base_config(/*dt=*/0.1, /*max_steps=*/5, /*goal_tolerance=*/0.05);
  StraightPlanner planner(/*v=*/0.5);  // base_config() sets stall_window > max_steps: no stall

  SimResult r = simulate(planner, grid, start, task, config, nullptr);

  ASSERT_EQ(r.status, SimStatus::TIMEOUT);
  EXPECT_FALSE(r.success);
  EXPECT_EQ(r.steps, config.max_steps);
  EXPECT_DOUBLE_EQ(r.time_to_goal, config.max_steps * config.control_dt);
}

// ⑤ constant (v, omega) traces the closed-form arc's circle ----------------------

TEST(LocalSim, ConstantArcCommandTracesExactCircle) {
  auto grid = open_grid(400, /*resolution=*/0.25);  // 100m x 100m: room for a 1m-radius loop
  const double v = 1.0, omega = 1.0;
  const double radius = v / omega;
  RobotState start{Pose{50.0, 50.0, 0.0}, 0.0, 0.0};
  // Turning left (omega > 0) from theta0=0: the arc's center is at (x0, y0 + radius).
  const double cx = 50.0, cy = 50.0 + radius;
  LocalTask task{Pose{-1000.0, -1000.0, 0.0}, {}};  // unreachable -> run the full budget
  SimConfig config = base_config(/*dt=*/0.05, /*max_steps=*/40, /*goal_tolerance=*/0.01);
  ConstantArcPlanner planner(v, omega);

  SimResult r = simulate(planner, grid, start, task, config, nullptr);

  ASSERT_EQ(r.status, SimStatus::TIMEOUT);
  for (const Pose& p : r.trajectory) {
    double d2 = (p.x - cx) * (p.x - cx) + (p.y - cy) * (p.y - cy);
    EXPECT_NEAR(std::sqrt(d2), radius, 1e-9);
  }
}

// ⑥ min_clearance / distance_traveled hand-check ---------------------------------

TEST(LocalSim, MinClearanceAndDistanceTraveledMatchHandComputation) {
  // 20m x 20m grid, resolution 1.0: one occupied cell, everything else free.
  // Occupied cell (row=10, col=15) sits at world center (15.5, 9.5) -- see
  // OccupancyGrid2D::cell_to_world.
  std::vector<bool> free_cells(20 * 20, true);
  free_cells[static_cast<size_t>(10) * 20 + 15] = false;
  maps::OccupancyGrid2D grid(20, 20, /*resolution=*/1.0, 0.0, 0.0, free_cells);

  // theta=0 keeps y exactly 9.5 (row 10) for the whole run -- distance_to_nearest
  // is a cell-center-to-cell-center query, so every tick's clearance is exactly
  // |15 - col| cells away from the obstacle at (row 10, col 15).
  RobotState start{Pose{10.5, 9.5, 0.0}, 0.0, 0.0};  // cell (row 10, col 10)
  LocalTask task{Pose{-1000.0, -1000.0, 0.0}, {}};    // unreachable -> run the full budget
  SimConfig config = base_config(/*dt=*/1.0, /*max_steps=*/3, /*goal_tolerance=*/0.01);
  StraightPlanner planner(/*v=*/1.0);  // 1 m/tick -> exactly one cell per tick

  SimResult r = simulate(planner, grid, start, task, config, nullptr);

  ASSERT_EQ(r.status, SimStatus::TIMEOUT);
  // clearance sequence (cells from col 15): start col10 -> 5, tick1 col11 -> 4,
  // tick2 col12 -> 3, tick3 col13 -> 2. Monotonically decreasing -> min is the
  // last tick's value.
  EXPECT_DOUBLE_EQ(r.min_clearance, 2.0);
  EXPECT_DOUBLE_EQ(r.distance_traveled, 3.0);  // 3 ticks * 1 m/tick, exact (sin/cos(0))
}

// ⑦ recorder emits the expected events; a null recorder emits nothing -----------

TEST(LocalSim, RecorderEmitsEventsAndNullRecorderIsSafe) {
  auto grid = open_grid(40);
  RobotState start{Pose{2.0, 2.0, 0.0}, 0.0, 0.0};  // clear of the map boundary
  LocalTask task{Pose{3.0, 2.0, 0.0}, {}};
  SimConfig config = base_config(/*dt=*/0.1, /*max_steps=*/50, /*goal_tolerance=*/0.05);

  StraightPlanner null_planner(0.5);
  EXPECT_NO_THROW(simulate(null_planner, grid, start, task, config, nullptr));

  StraightPlanner traced_planner(0.5);
  std::ostringstream os;
  TraceRecorder rec(os);
  SimResult r = simulate(traced_planner, grid, start, task, config, &rec);
  ASSERT_TRUE(r.success);

  std::string trace = os.str();
  EXPECT_NE(trace.find("\"robot_moved\""), std::string::npos);
  EXPECT_NE(trace.find("\"path_found\""), std::string::npos);
  EXPECT_NE(trace.find("\"planning_finished\""), std::string::npos);
}

// ⑧ reset() is called at episode start: re-running the same instance is deterministic

TEST(LocalSim, ResetMakesRerunOfSameInstanceIdentical) {
  auto grid = open_grid(60);
  RobotState start{Pose{5.0, 5.0, 0.0}, 0.0, 0.0};
  LocalTask task{Pose{8.0, 6.0, 0.0}, {}};
  SimConfig config = base_config(/*dt=*/0.1, /*max_steps=*/60, /*goal_tolerance=*/0.05);
  CountingPlanner planner(/*turn_ticks=*/4);

  SimResult r1 = simulate(planner, grid, start, task, config, nullptr);
  SimResult r2 = simulate(planner, grid, start, task, config, nullptr);

  EXPECT_EQ(r1.status, r2.status);
  EXPECT_EQ(r1.steps, r2.steps);
  EXPECT_DOUBLE_EQ(r1.distance_traveled, r2.distance_traveled);
  ASSERT_EQ(r1.trajectory.size(), r2.trajectory.size());
  for (size_t i = 0; i < r1.trajectory.size(); ++i) {
    EXPECT_DOUBLE_EQ(r1.trajectory[i].x, r2.trajectory[i].x);
    EXPECT_DOUBLE_EQ(r1.trajectory[i].y, r2.trajectory[i].y);
    EXPECT_DOUBLE_EQ(r1.trajectory[i].theta, r2.trajectory[i].theta);
  }
}

// integrate_unicycle: straight branch (omega ~ 0) -------------------------------

TEST(LocalSim, IntegrateUnicycleStraightBranch) {
  Pose p = local_planning::integrate_unicycle(Pose{0.0, 0.0, 0.0}, VelocityCommand{2.0, 0.0}, 0.5);
  EXPECT_DOUBLE_EQ(p.x, 1.0);
  EXPECT_DOUBLE_EQ(p.y, 0.0);
  EXPECT_DOUBLE_EQ(p.theta, 0.0);
}

// integrate_unicycle: exact-arc branch matches the closed form directly --------

TEST(LocalSim, IntegrateUnicycleArcBranchMatchesClosedForm) {
  double v = 1.0, omega = 0.5, dt = 0.2, theta0 = 0.3;
  Pose p = local_planning::integrate_unicycle(Pose{1.0, 2.0, theta0}, VelocityCommand{v, omega}, dt);
  double theta1 = theta0 + omega * dt;
  double expected_x = 1.0 + (v / omega) * (std::sin(theta1) - std::sin(theta0));
  double expected_y = 2.0 - (v / omega) * (std::cos(theta1) - std::cos(theta0));
  EXPECT_NEAR(p.x, expected_x, 1e-12);
  EXPECT_NEAR(p.y, expected_y, 1e-12);
  EXPECT_NEAR(p.theta, local_planning::wrap_to_pi(theta1), 1e-12);
}

File: python/tests/test_local_sim.py
"""Closed-loop simulator behavior: termination modes, integration, trace contract."""

from __future__ import annotations

import io
import json
import math

import pytest
from conftest import grid_from, open_grid

from navigation.core.capabilities import Capability, ObstacleQuery
from navigation.core.params import ParamSet
from navigation.core.planner import LocalPlanner
from navigation.core.trace import TraceRecorder
from navigation.core.types import LocalTask, RobotState, VelocityCommand
from navigation.local_planning._geometry import wrap_to_pi
from navigation.local_planning.simulation import SimConfig, SimStatus, integrate_unicycle, simulate


def _script_params() -> ParamSet:
    # Script planners below declare no algorithm params of their own -- the
    # empty ParamSet just satisfies LocalPlanner's constructor contract.
    return ParamSet("script", "local_planning", {})


class _ScriptPlanner(LocalPlanner[ObstacleQuery]):
    """Shared base for the test-only fixture planners below.

    Not a mock: each subclass drives a real tick loop through `simulate()` to
    exercise the simulator's integration / termination / metric behavior, the
    thing this test module verifies.
    """

    def __init__(self) -> None:
        super().__init__(_script_params())

    def required_capabilities(self) -> set[Capability]:
        return {Capability.OBSTACLE_QUERY}


class _StraightPlanner(_ScriptPlanner):
    """Constant (v, omega) every tick -- the plain no-obstacle-awareness case."""

    def __init__(self, v: float = 1.0, omega: float = 0.0) -> None:
        super().__init__()
        self._v = v
        self._omega = omega

    @property
    def name(self) -> str:
        return "script_straight"

    def compute_command(
        self,
        space: ObstacleQuery,
        state: RobotState,
        task: LocalTask,
        dt: float,
        recorder: TraceRecorder | None = None,
    ) -> VelocityCommand:
        return VelocityCommand(self._v, self._omega)


class _ZeroPlanner(_ScriptPlanner):
    """Always issues the zero command -- drives the stall detector."""

    @property
    def name(self) -> str:
        return "script_zero"

    def compute_command(
        self,
        space: ObstacleQuery,
        state: RobotState,
        task: LocalTask,
        dt: float,
        recorder: TraceRecorder | None = None,
    ) -> VelocityCommand:
        return VelocityCommand(0.0, 0.0)


class _RampPlanner(_ScriptPlanner):
    """Drives forward for a fixed number of ticks, then stops.

    Its command depends on an internal tick counter, so a run only replays
    identically if `simulate()` actually calls `reset()` at the start of
    every episode -- this is what the reset-contract test below checks.
    """

    def __init__(self, forward_ticks: int, v: float = 1.0) -> None:
        super().__init__()
        self._forward_ticks = forward_ticks
        self._v = v
        self._tick = 0

    @property
    def name(self) -> str:
        return "script_ramp"

    def reset(self) -> None:
        self._tick = 0

    def compute_command(
        self,
        space: ObstacleQuery,
        state: RobotState,
        task: LocalTask,
        dt: float,
        recorder: TraceRecorder | None = None,
    ) -> VelocityCommand:
        v = self._v if self._tick < self._forward_ticks else 0.0
        self._tick += 1
        return VelocityCommand(v, 0.0)


# --- 1. straight-line reach -------------------------------------------------
def test_straight_line_reaches_goal_with_exact_time_to_goal() -> None:
    grid = open_grid(5, 5)
    planner = _StraightPlanner(v=1.0, omega=0.0)
    start = RobotState(pose=(0.5, 2.5, 0.0))
    task = LocalTask(goal=(4.5, 2.5, 0.0))
    config = SimConfig(
        control_dt=0.5, max_steps=20, goal_tolerance=0.3,
        footprint_radius=0.1, stall_window=100, stall_distance=0.0,
    )
    result = simulate(planner, grid, start, task, config)
    assert result.status is SimStatus.REACHED
    assert result.success is True
    assert result.steps == 8
    assert result.time_to_goal == pytest.approx(8 * 0.5)


# --- 2. collision ------------------------------------------------------------
def test_start_pose_already_colliding_is_immediate_collision() -> None:
    grid = grid_from(["###", "###", "###"])
    planner = _StraightPlanner()
    start = RobotState(pose=(1.5, 1.5, 0.0))
    task = LocalTask(goal=(1.5, 1.5, 0.0))
    config = SimConfig(
        control_dt=0.1, max_steps=10, goal_tolerance=0.1,
        footprint_radius=0.1, stall_window=5, stall_distance=0.01,
    )
    result = simulate(planner, grid, start, task, config)
    assert result.status is SimStatus.COLLISION
    assert result.success is False
    assert result.steps == 0
    assert result.trajectory == [start.pose]


def test_wall_ahead_collides_mid_loop() -> None:
    # Single obstacle at cell (2,2) -> world center (2.5, 2.5); the robot drives
    # straight into it along row 2 (also world y=2.5), a branch distinct from
    # the immediate-start-collision case above (different code path entirely).
    grid = grid_from(
        [
            ".....",
            ".....",
            "..#..",
            ".....",
            ".....",
        ]
    )
    planner = _StraightPlanner(v=1.0, omega=0.0)
    start = RobotState(pose=(0.5, 2.5, 0.0))
    task = LocalTask(goal=(4.5, 2.5, 0.0))
    config = SimConfig(
        control_dt=0.5, max_steps=10, goal_tolerance=0.05,
        footprint_radius=0.2, stall_window=100, stall_distance=0.0,
    )
    result = simulate(planner, grid, start, task, config)
    assert result.status is SimStatus.COLLISION
    assert result.success is False
    assert 0 < result.steps < 10


# --- 3. stall ------------------------------------------------------------
def test_zero_command_triggers_stall() -> None:
    grid = open_grid(5, 5)
    planner = _ZeroPlanner()
    start = RobotState(pose=(2.5, 2.5, 0.0))
    task = LocalTask(goal=(4.9, 4.9, 0.0))
    config = SimConfig(
        control_dt=0.1, max_steps=20, goal_tolerance=0.05,
        footprint_radius=0.1, stall_window=5, stall_distance=0.02,
    )
    result = simulate(planner, grid, start, task, config)
    assert result.status is SimStatus.STALLED
    assert result.success is False
    assert result.steps == 5
    assert result.distance_traveled == 0.0


# --- 4. timeout ------------------------------------------------------------
def test_unreachable_goal_times_out() -> None:
    grid = open_grid(5, 5)
    planner = _StraightPlanner(v=1.0, omega=0.0)
    start = RobotState(pose=(2.5, 2.5, 0.0))
    task = LocalTask(goal=(-100.0, -100.0, 0.0))  # behind the robot's forward heading
    config = SimConfig(
        control_dt=0.1, max_steps=6, goal_tolerance=0.05,
        footprint_radius=0.1, stall_window=100, stall_distance=0.0,
    )
    result = simulate(planner, grid, start, task, config)
    assert result.status is SimStatus.TIMEOUT
    assert result.success is False
    assert result.steps == 6


# --- 5. exact-arc integration ------------------------------------------------
def test_integrate_unicycle_constant_command_traces_a_circle() -> None:
    v, omega = 1.0, 0.5
    r = v / omega
    pose = (0.0, 0.0, 0.0)
    cmd = VelocityCommand(v, omega)
    dt = 0.05
    # (x - r*sin(theta), y + r*cos(theta)) is the invariant circle center for a
    # constant-(v, omega) unicycle arc; every subsequent pose must sit at
    # radius r from it if the closed-form integration is exact.
    center = (pose[0] - r * math.sin(pose[2]), pose[1] + r * math.cos(pose[2]))
    for _ in range(40):
        pose = integrate_unicycle(pose, cmd, dt)
        radius = math.hypot(pose[0] - center[0], pose[1] - center[1])
        assert radius == pytest.approx(r, abs=1e-9)


# --- wrap_to_pi: both normalization branches (integrate_unicycle's turns stay
# small in the tests above, so this covers the negative-angle wraparound too) --
def test_wrap_to_pi_normalizes_both_directions() -> None:
    assert wrap_to_pi(0.5) == pytest.approx(0.5)
    assert wrap_to_pi(math.pi) == pytest.approx(math.pi)
    assert wrap_to_pi(-math.pi) == pytest.approx(math.pi)  # (-pi, pi]: -pi wraps to +pi
    assert wrap_to_pi(3.0 * math.pi) == pytest.approx(math.pi)
    assert wrap_to_pi(-3.0 * math.pi) == pytest.approx(math.pi)


# --- 6. min_clearance / distance_traveled hand-check -------------------------
def test_min_clearance_and_distance_traveled_match_hand_computation() -> None:
    grid = open_grid(5, 5)
    planner = _StraightPlanner(v=1.0, omega=0.0)
    start = RobotState(pose=(0.5, 2.5, 0.0))
    task = LocalTask(goal=(100.0, 100.0, 0.0))  # far enough that REACHED never fires
    config = SimConfig(
        control_dt=0.2, max_steps=5, goal_tolerance=0.01,
        footprint_radius=0.1, stall_window=100, stall_distance=0.0,
    )
    result = simulate(planner, grid, start, task, config)
    assert result.status is SimStatus.TIMEOUT

    xs = [0.5 + 0.2 * i for i in range(6)]  # start + 5 ticks at v*dt = 0.2 each
    expected_clearance = min(grid.distance_to_nearest((x, 2.5)) for x in xs)
    assert result.min_clearance == pytest.approx(expected_clearance)
    assert result.distance_traveled == pytest.approx(0.2 * 5)


# --- 7. trace contract + zero-cost None recorder ------------------------------
def test_recorder_emits_events_and_none_recorder_is_zero_cost() -> None:
    grid = open_grid(5, 5)
    planner = _StraightPlanner(v=1.0, omega=0.0)
    start = RobotState(pose=(0.5, 2.5, 0.0))
    task = LocalTask(goal=(1.5, 2.5, 0.0))
    config = SimConfig(
        control_dt=0.5, max_steps=10, goal_tolerance=0.1,
        footprint_radius=0.1, stall_window=100, stall_distance=0.0,
    )

    buf = io.StringIO()
    recorder = TraceRecorder(buf)
    result_with = simulate(planner, grid, start, task, config, recorder)
    events = [json.loads(line)["event"] for line in buf.getvalue().splitlines()]
    assert result_with.status is SimStatus.REACHED
    assert events.count("robot_moved") == result_with.steps
    assert events.count("path_found") == 1
    assert events.count("planning_finished") == 1

    result_without = simulate(planner, grid, start, task, config, None)
    assert result_without.status == result_with.status
    assert result_without.steps == result_with.steps
    assert result_without.distance_traveled == pytest.approx(result_with.distance_traveled)


# --- 8. reset contract ------------------------------------------------------
def test_reset_makes_reruns_deterministic() -> None:
    grid = open_grid(5, 5)
    planner = _RampPlanner(forward_ticks=3, v=1.0)
    start = RobotState(pose=(0.5, 2.5, 0.0))
    task = LocalTask(goal=(100.0, 100.0, 0.0))  # unreachable -> outcome is script-driven
    config = SimConfig(
        control_dt=0.2, max_steps=10, goal_tolerance=0.01,
        footprint_radius=0.1, stall_window=3, stall_distance=0.01,
    )

    first = simulate(planner, grid, start, task, config)
    second = simulate(planner, grid, start, task, config)

    assert second.status == first.status
    assert second.steps == first.steps
    assert second.distance_traveled == pytest.approx(first.distance_traveled)
